packages/react-router/index.d.ts

```diff
@@ -0,0 +1,179 @@
+}
+
+/**
+ * Navigate programmatically using a component.
+ */
+export class Navigate extends React.Component<NavigateProps, any> {
+}
+
+/**
+ * Renders the child route's element, if there is one.
+ */
+export class Outlet extends React.Component<{}, any> {
+}
+
+export interface RouteProps {
+  path?: string;
+  element?: React.ReactElement;
+}
+
+/**
+ * Used in a route config to render an element.
+ */
+export class Route extends React.Component<RouteProps, any> {
+}
+
+export interface RouterProps {
+  history?: History;
+  timeout?: number;
+}
+
+/**
+ * The root context provider. There should be only one of these in a given app.
+ */
+export class Router extends React.Component<RouterProps, any> {
+}
+
+export interface RoutesProps {
+  basename?: string;
+  caseSensitive?: boolean;
+}
+
+/**
+ * A wrapper for useRoutes that treats its children as route and/or redirect
+ * objects.
+ */
+export class Routes extends React.Component<RoutesProps, any> {
+}
+
+/**
+ * Utility function that creates a routes config object from a React
+ * "children" object, which is usually either a React element or an
+ * array of elements.
+ */
+export function createRoutesFromChildren(children: React.ReactNode): RouteConfigObject[];
+
+/**
+ * Blocks all navigation attempts. This is useful for preventing the page from
+ * changing until some condition is met, like saving form data.
+ */
+export function useBlocker(blocker: any, when?: boolean): void;
+
+/**
+ * Returns the full href for the given "to" value. This is useful for building
+ * custom links that are also accessible and preserve right-click behavior.
+ */
+export function useHref(to: History.LocationDescriptor): History.Href;
+
+/**
+ * Returns true if this component is a descendant of a <Router>.
+ */
+export function useInRouterContext(): boolean;
+
+/**
+ * Returns the current location object, which represents the current URL in web
+ * browsers.
+ *
+ * NOTE: If you're using this it may mean you're doing some of your own "routing"
+ * in your app, and we'd like to know what your use case is. We may be able to
+ * provide something higher-level to better suit your needs.
+ */
+export function useLocation(): History.Location;
+
+/**
+ * Returns true if the router is pending a location update.
+ */
+export function useLocationPending(): boolean;
+
+/**
+ * Returns true if the URL for the given "to" value matches the current URL.
+ * This is useful for components that need to know "active" state, e.g.
+ * <NavLink>.
+ */
+export function useMatch(to: History.Location): boolean;
+
+export interface NavigateOptions {
+  replace?: boolean;
+  state?: any;
+}
+
+declare function navigate(delta: number): void;
+declare function navigate(to: History.LocationDescriptor, options?: NavigateOptions): void;
+
+/**
+ * Returns an imperative method for changing the location. Used by <Link>s, but
+ * may also be used by other elements to change the location.
+ */
+export function useNavigate(): typeof navigate;
```

Comment:
this is wip, just want to suggest some 
`type navigateFn<T> = (to: string | number, options?: NavigateOptions<T>) => void
export function useNavigate<T = Record<string, any>> (to: string | number, options?: Omit<NavigateProps<T>, 'to'>): navigateFn<T>;`

Reply:
The problem with `type navigateFn` you suggest is that it indicates that `navigate` has a call signature `navigate(to: number, options?)` which is actually not correct, according to the implementation. That's why I prefer `declare function`.

Adding generics support is a good idea. 😄